lib/logscheduler.c

```diff
@@ -0,0 +1,429 @@
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "logscheduler.h"
+#include "scratch-buffers.h"
+
+static void
+_reinject_message(LogPipe *front_pipe, LogMessage *msg, const LogPathOptions *path_options)
+{
+  if (front_pipe)
+    log_pipe_queue(front_pipe, msg, path_options);
+  else
+    log_msg_drop(msg, path_options, AT_PROCESSED);
+}
+
+#if SYSLOG_NG_HAVE_IV_WORK_POOL_SUBMIT_CONTINUATION
+
+/* LogSchedulerBatch */
+
+LogSchedulerBatch *
+_batch_new(struct iv_list_head *elements)
+{
+  LogSchedulerBatch *batch = g_new0(LogSchedulerBatch, 1);
+
+  INIT_IV_LIST_HEAD(&batch->elements);
+  INIT_IV_LIST_HEAD(&batch->list);
+  iv_list_splice_tail(elements, &batch->elements);
+  return batch;
+}
+
+void
+_batch_free(LogSchedulerBatch *batch)
+{
+  g_free(batch);
+}
+
+/* LogSchedulerPartition */
+
+static void
+_work(gpointer s, gpointer arg)
+{
+  LogSchedulerPartition *partition = (LogSchedulerPartition *) s;
+  struct iv_list_head *ilh, *next;
+  struct iv_list_head *ilh2, *next2;
+  gint num = 0;
+
+  /* batches_lock protects the batches list itself.  We take off partitions
+   * one-by-one under the protection of the lock */
+
+  g_mutex_lock(&partition->batches_lock);
+  while (!iv_list_empty(&partition->batches))
+    {
+      struct iv_list_head batches = IV_LIST_HEAD_INIT(batches);
+      iv_list_splice_init(&partition->batches, &batches);
+
+      g_mutex_unlock(&partition->batches_lock);
+
+      iv_list_for_each_safe(ilh, next, &batches)
+      {
+        /* remove the first batch from the batches list */
+        LogSchedulerBatch *batch = iv_list_entry(ilh, LogSchedulerBatch, list);
+        iv_list_del(&batch->list);
+
+        iv_list_for_each_safe(ilh2, next2, &batch->elements)
+        {
+          LogMessageQueueNode *node = iv_list_entry(ilh2, LogMessageQueueNode, list);
+
+          iv_list_del(&node->list);
+
+          LogMessage *msg = log_msg_ref(node->msg);
+
+          LogPathOptions path_options = LOG_PATH_OPTIONS_INIT;
+          path_options.ack_needed = node->ack_needed;
+          path_options.flow_control_requested = node->flow_control_requested;
+
+          log_msg_free_queue_node(node);
+
+          log_msg_refcache_start_consumer(msg, &path_options);
```

Comment:
Why do we call the consumer version of the refcache here?

Reply:
The LogScheduler is not a producer of log messages, so we can't use that. While there can only be one producer, there can be multiple consumers.

maybe the naming for consumer mode is not perfect. it applies to anything that takes over messages from another thread for processing.